lib/Conversion/GPUXToLLVM/GPUXToLLVMPass.cpp

```diff
@@ -0,0 +1,676 @@
+          llvmInt32Type    // shared
+      }};
+
+  FunctionCallBuilder deallocCallBuilder = {
+      "gpuMemFree",
+      llvmVoidType,
+      {
+          llvmPointerType, /* void *stream */
+          llvmPointerType  /* void *ptr */
+      }};
+};
+
+/// A rewrite pattern to convert gpux.alloc operations into a GPU runtime
+/// call.
+class ConvertAllocOpToGpuRuntimeCallPattern
+    : public ConvertOpToGpuRuntimeCallPattern<gpux::AllocOp> {
+public:
+  ConvertAllocOpToGpuRuntimeCallPattern(mlir::LLVMTypeConverter &typeConverter)
+      : ConvertOpToGpuRuntimeCallPattern<gpux::AllocOp>(typeConverter) {}
+
+private:
+  mlir::LogicalResult
+  matchAndRewrite(gpux::AllocOp allocOp, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+    if (!allocOp.getSymbolOperands().empty())
+      return mlir::failure();
+
+    mlir::MemRefType memRefType = allocOp.getType();
+    auto converter = getTypeConverter();
+    auto dstType = converter->convertType(memRefType);
+    if (!dstType)
+      return mlir::failure();
+
+    bool isShared = allocOp.getHostShared();
+
+    auto loc = allocOp.getLoc();
+
+    // Get shape of the memref as values: static sizes are constant
+    // values and dynamic sizes are passed to 'alloc' as operands.
+    mlir::SmallVector<mlir::Value, 4> shape;
+    mlir::SmallVector<mlir::Value, 4> strides;
+    mlir::Value sizeBytes;
+    getMemRefDescriptorSizes(loc, memRefType, adaptor.getDynamicSizes(),
+                             rewriter, shape, strides, sizeBytes);
+
+    assert(shape.size() == strides.size());
+
+    auto alignment = rewriter.getIntegerAttr(llvmIndexType, 64);
+    auto alignmentVar =
+        rewriter.create<mlir::LLVM::ConstantOp>(loc, llvmIndexType, alignment);
+
+    auto typeVar = rewriter.create<mlir::LLVM::ConstantOp>(
+        loc, llvmInt32Type, rewriter.getI32IntegerAttr(isShared));
+
+    mlir::Value allocatedPtr =
+        allocCallBuilder
+            .create(loc, rewriter,
+                    {adaptor.getGpuxStream(), sizeBytes, alignmentVar, typeVar})
+            ->getResult(0);
+
+    auto memrefDesc = mlir::MemRefDescriptor::undef(rewriter, loc, dstType);
+    auto elemPtrTye = memrefDesc.getElementPtrType();
+    memrefDesc.setAllocatedPtr(
+        rewriter, loc,
+        rewriter.create<mlir::LLVM::BitcastOp>(loc, elemPtrTye, allocatedPtr));
+    memrefDesc.setAlignedPtr(
+        rewriter, loc,
+        rewriter.create<mlir::LLVM::BitcastOp>(loc, elemPtrTye, allocatedPtr));
+
+    auto zero = rewriter.create<mlir::LLVM::ConstantOp>(
+        loc, llvmIndexType, rewriter.getIntegerAttr(llvmIndexType, 0));
+
+    memrefDesc.setOffset(rewriter, loc, zero);
+    for (auto i : llvm::seq(0u, static_cast<unsigned>(shape.size()))) {
+      memrefDesc.setSize(rewriter, loc, i, shape[i]);
+      memrefDesc.setStride(rewriter, loc, i, strides[i]);
+    }
+
+    mlir::Value resMemref = memrefDesc;
+    rewriter.replaceOp(allocOp, resMemref);
```

Comment:
How should I understand allocOp is replaced by a "mlir::Value"?  I thought we just create  MLIR LLVM dialect op calling the wrapper function like gpuMemAlloc